File: netlify/functions/crosscheck.py
from concurrent.futures import ThreadPoolExecutor

from credify_common import PROVIDERS, json_response, judge, parse_body


def _review(name, prompt, deep):
    try:
        return PROVIDERS[name](prompt, deep)
    except Exception:
        return None


def handler(event, context):
    data = parse_body(event)
    question = (data.get("question") or "").strip()
    deep = bool(data.get("deep"))
    answers = data.get("answers") or {}
    claude_answer, chatgpt_answer = answers.get("claude"), answers.get("chatgpt")
    if not question or not claude_answer or not chatgpt_answer:
        return json_response(400, {"error": "Need both models' answers before cross-checking."})

    claude_prompt = (
        f"Question/text: {question}\n\nYour previous answer: {claude_answer}\n\n"
        f"Another AI (ChatGPT) answered: {chatgpt_answer}\n\n"
        "Review both answers. If ChatGPT caught something you missed, or you "
        "spot an error in either answer, correct it. Give your refined final "
        "answer only, no preamble."
    )
    chatgpt_prompt = (
        f"Question/text: {question}\n\nYour previous answer: {chatgpt_answer}\n\n"
        f"Another AI (Claude) answered: {claude_answer}\n\n"
        "Review both answers. If Claude caught something you missed, or you "
        "spot an error in either answer, correct it. Give your refined final "
        "answer only, no preamble."
    )

    with ThreadPoolExecutor(max_workers=2) as pool:
        f_claude = pool.submit(_review, "claude", claude_prompt, deep)
        f_chatgpt = pool.submit(_review, "chatgpt", chatgpt_prompt, deep)
        refined_claude = f_claude.result()
        refined_chatgpt = f_chatgpt.result()

    final_claude = refined_claude or claude_answer
    final_chatgpt = refined_chatgpt or chatgpt_answer
    verdict = judge(question, {"claude": final_claude, "chatgpt": final_chatgpt})

    resp = {
        "claude": refined_claude,
        "chatgpt": refined_chatgpt,
        "note": "each model reviewed the other's answer and gave its final version above.",
    }
    if verdict:
        resp["agreement"] = {
            "score": verdict["score"],
            "word": verdict["word"],
            "bottom": verdict["bottom"],
        }
    return json_response(200, resp)

File: netlify/functions/crosscheck.js
import { PROVIDERS, judge } from "./lib/providers.js";

const json = (status, payload) =>
  new Response(JSON.stringify(payload), { status, headers: { "content-type": "application/json" } });

async function review(name, prompt, deep) {
  try {
    return await PROVIDERS[name](prompt, deep);
  } catch {
    return null;
  }
}

export default async (req) => {
  if (req.method !== "POST") return json(405, { error: "Method not allowed." });

  let data;
  try {
    data = await req.json();
  } catch {
    data = {};
  }
  const question = (data.question || "").trim();
  const deep = Boolean(data.deep);
  const answers = data.answers || {};
  const claudeAnswer = answers.claude;
  const chatgptAnswer = answers.chatgpt;
  if (!question || !claudeAnswer || !chatgptAnswer) {
    return json(400, { error: "Need both models' answers before cross-checking." });
  }

  const claudePrompt =
    `Question/text: ${question}\n\nYour previous answer: ${claudeAnswer}\n\n` +
    `Another AI (ChatGPT) answered: ${chatgptAnswer}\n\n` +
    "Review both answers. If ChatGPT caught something you missed, or you " +
    "spot an error in either answer, correct it. Give your refined final " +
    "answer only, no preamble.";
  const chatgptPrompt =
    `Question/text: ${question}\n\nYour previous answer: ${chatgptAnswer}\n\n` +
    `Another AI (Claude) answered: ${claudeAnswer}\n\n` +
    "Review both answers. If Claude caught something you missed, or you " +
    "spot an error in either answer, correct it. Give your refined final " +
    "answer only, no preamble.";

  const [refinedClaude, refinedChatgpt] = await Promise.all([
    review("claude", claudePrompt, deep),
    review("chatgpt", chatgptPrompt, deep),
  ]);

  const finalClaude = refinedClaude || claudeAnswer;
  const finalChatgpt = refinedChatgpt || chatgptAnswer;
  const verdict = await judge(question, { claude: finalClaude, chatgpt: finalChatgpt });

  const resp = {
    claude: refinedClaude,
    chatgpt: refinedChatgpt,
    note: "each model reviewed the other's answer and gave its final version above.",
  };
  if (verdict) {
    resp.agreement = { score: verdict.score, word: verdict.word, bottom: verdict.bottom };
  }
  return json(200, resp);
};

export const config = { path: "/api/crosscheck" };
